# HG changeset patch
# User Michael Goffioul <michael.goffioul@gmail.com>
# Date 1266775499 0
#      Sun Feb 21 18:04:59 2010 +0000
# Node ID 65d5776379c3b1bd054be8ab5f1ac25cfa9ebe3e
# Parent  33b012d39dce14d975b651ae62e1ed0e2516afad
Reduce the amount of stuff included by windows.h and avoid min/max being #define-d

diff --git a/ChangeLog b/ChangeLog
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,10 +1,12 @@
 2010-02-21  Michael Goffioul  <michael.goffioul@gmail.com>
 
+	* configure.ac: Define NOMINMAX under MSVC.
+
 	* configure.ac: Add pcre_config --cflags to XTRA_CXXFLAGS instead of
 	CPPFLAGS.
 
 2010-02-17  John W. Eaton  <jwe@octave.org>
 
 	* configure.ac: Add $COLAMD_LDFLAGS $COLAMD_LIBS $AMD_LDFLAGS,
 	and $AMD_LIBS to LIBS when checking for cholmod library.
 
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -425,16 +425,21 @@ case "$canonical_host_type" in
     OCTAVE_CXX_FLAG(-mminimal-toc, [
       XTRA_CXXFLAGS="$XTRA_CXXFLAGS -mminimal-toc"])
   ;;
 esac
 
 AC_SUBST(XTRA_CFLAGS)
 AC_SUBST(XTRA_CXXFLAGS)
 
+## Avoid #define of min/max from windows.h header
+if test "$have_msvc" = "yes"; then
+  AC_DEFINE(NOMINMAX, 1, [Define if you want to avoid min/max macro definition in Windows headers])
+fi
+
 ### Use -static if compiling on Alpha OSF/1 1.3 systems.
 
 case "$canonical_host_type" in
   alpha*-dec-osf1.3)
     LD_STATIC_FLAG=-static
   ;;
 esac
 if test -n "$LD_STATIC_FLAG"; then
diff --git a/libcruft/ChangeLog b/libcruft/ChangeLog
--- a/libcruft/ChangeLog
+++ b/libcruft/ChangeLog
@@ -1,10 +1,12 @@
 2010-02-21  Michael Goffioul  <michael.goffioul@gmail.com>
 
+	* misc/quit.h: Define WIN32_LEAN_AND_MEAN and don't #undef min/max.
+
 	* Makefile.am: Add -bindir option to libcruft_la_LDFLAGS.
 
 2010-02-17  John W. Eaton  <jwe@octave.org>
 
 	* Makefile.am (libcruft_la_LIBADD): Include libranlib.la in the list.
 	(noinst_LTLIBRARIES, libranlib_la_SOURCES, libranlib_la_DEPENDENCIES,
 	libcruft_la_FFLAGS): New variables.
 	(ranlib.def): New target.
diff --git a/libcruft/misc/quit.h b/libcruft/misc/quit.h
--- a/libcruft/misc/quit.h
+++ b/libcruft/misc/quit.h
@@ -30,19 +30,18 @@ extern "C" {
 
 #include <stdio.h>
 
 #include <signal.h>
 #include <setjmp.h>
 
 #if defined (__WIN32__) && ! defined (_POSIX_VERSION)
 
+#define WIN32_LEAN_AND_MEAN
 #include <windows.h>
-#undef min
-#undef max
 
 CRUFT_API extern void w32_sigint_init (void);   /* setup */
 CRUFT_API extern void w32_raise_final (void);   /* tear down */
 CRUFT_API extern void w32_raise (int sig);      /* raise signal in main thread */
 CRUFT_API extern int w32_in_main_thread (void); /* return true if in main thread */
 
 #endif
 
diff --git a/liboctave/ChangeLog b/liboctave/ChangeLog
--- a/liboctave/ChangeLog
+++ b/liboctave/ChangeLog
@@ -1,10 +1,14 @@
 2010-02-21  Michael Goffioul  <michael.goffioul@gmail.com>
 
+	* lo-sysdep.cc, lo-utils.h, oct-shlib.h: Define WIN32_LEAN_AND_MEAN
+	and don't #undef min/max.
+	* oct-inttypes.h: Don't #undef min/max.
+
 	* oct-glob.cc (octave_glob): Convert backslashes to forward slashes
 	automatically before calling glob, and convert back after the call
 	(implementation from jwe).
 
 	* Makefile.am: Add -bindir flag to liboctave_la_LDFLAGS.
 
 2010-02-19  Jaroslav Hajek  <highegg@gmail.com>
 
diff --git a/liboctave/lo-sysdep.cc b/liboctave/lo-sysdep.cc
--- a/liboctave/lo-sysdep.cc
+++ b/liboctave/lo-sysdep.cc
@@ -29,21 +29,18 @@ along with Octave; see the file COPYING.
 #include <vector>
 
 #include <sys/types.h>
 #include <unistd.h>
 
 #include <fcntl.h>
 
 #if defined (__WIN32__) && ! defined (__CYGWIN__)
+#define WIN32_LEAN_AND_MEAN
 #include <windows.h>
-#ifdef max
-# undef min
-# undef max
-#endif
 #endif
 
 #include "file-ops.h"
 #include "lo-error.h"
 #include "pathlen.h"
 #include "lo-sysdep.h"
 #include "str-vec.h"
 #include "oct-locbuf.h"
diff --git a/liboctave/lo-utils.h b/liboctave/lo-utils.h
--- a/liboctave/lo-utils.h
+++ b/liboctave/lo-utils.h
@@ -109,18 +109,18 @@ octave_write_complex (std::ostream& os, 
 
 extern OCTAVE_API void
 octave_write_float (std::ostream& os, float dval);
 
 extern OCTAVE_API void
 octave_write_float_complex (std::ostream& os, const FloatComplex& cval);
 
 #ifdef HAVE_LOADLIBRARY_API
+#define WIN32_LEAN_AND_MEAN
 #include <windows.h>
 extern "C" OCTAVE_API void *
 octave_w32_library_search (HINSTANCE handle, const char *name);
-#undef min
-#undef max
-#endif
 #endif
 
 extern "C" OCTAVE_API pid_t
 octave_waitpid (pid_t pid, int *status, int options);
+
+#endif
diff --git a/liboctave/oct-inttypes.h b/liboctave/oct-inttypes.h
--- a/liboctave/oct-inttypes.h
+++ b/liboctave/oct-inttypes.h
@@ -34,22 +34,16 @@ along with Octave; see the file COPYING.
 #include "lo-math.h"
 #include "lo-mappers.h"
 
 #ifdef OCTAVE_INT_USE_LONG_DOUBLE
 inline long double xround (long double x) { return roundl (x); }
 inline long double xisnan (long double x) { return xisnan (static_cast<double> (x)); }
 #endif
 
-// Undefine min/max if needed (this may happen under Windows)
-#ifdef min
-#undef min
-#undef max
-#endif
-
 // FIXME -- we define this by our own because some compilers, such as
 // MSVC, do not provide std::abs (int64_t) and std::abs (uint64_t).  In
 // the future, it should go away in favor of std::abs.
 template <class T>
 inline T octave_int_abs (T x) { return x >= 0 ? x : -x; }
 
 // Query for an integer type of certain sizeof, and signedness.
 template<int qsize, bool qsigned>
diff --git a/liboctave/oct-shlib.cc b/liboctave/oct-shlib.cc
--- a/liboctave/oct-shlib.cc
+++ b/liboctave/oct-shlib.cc
@@ -45,19 +45,18 @@ extern "C"
 extern void *dlopen (const char *, int);
 extern const char *dlerror (void);
 extern void *dlsym (void *, const char *);
 extern int dlclose (void *);
 #endif
 #elif defined (HAVE_SHL_LOAD_API)
 #include <dl.h>
 #elif defined (HAVE_LOADLIBRARY_API)
+#define WIN32_LEAN_AND_MEAN
 #include <windows.h>
-#undef min
-#undef max
 #endif
 }
 
 #include "file-stat.h"
 #include "lo-error.h"
 #include "oct-shlib.h"
 #include "str-vec.h"
 
diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,10 +1,13 @@
 2010-02-21  Michael Goffioul  <michael.goffioul@gmail.com>
 
+	* gl-render.h: Define WIN32_LEAN_AND_MEAN and don't #undef min/max.
+	* DLD-FUNCTIONS/fltk_backend.cc: Don't #undef min/max.
+
 	* sighandlers.cc (octave_set_signal_handler): Don't use SIGALRM
 	unconditionally.
 
 	* ov-bool.h (class octave_bool): Add OCTINTERP_API.
 	* unwind-prot.h (class unwind_protect_safe): Ditto.
 
 	* Makefile.am: Add -bindir flag to liboctinterp_la_LDFLAGS. Remove
 	temporary libtool module file after copying oct file.
diff --git a/src/DLD-FUNCTIONS/fltk_backend.cc b/src/DLD-FUNCTIONS/fltk_backend.cc
--- a/src/DLD-FUNCTIONS/fltk_backend.cc
+++ b/src/DLD-FUNCTIONS/fltk_backend.cc
@@ -44,21 +44,16 @@ To initialize:
 #include <FL/Fl_Window.H>
 #include <FL/Fl_Output.H>
 #include <FL/Fl_Button.H>
 #include <FL/Fl_Gl_Window.H>
 #include <FL/fl_ask.H>
 #include <FL/fl_draw.H>
 #include <FL/gl.h>
 
-#ifdef min
-#undef min
-#undef max
-#endif
-
 #include "cmd-edit.h"
 #include "lo-ieee.h"
 
 #include "defun-dld.h"
 #include "error.h"
 #include "gl-render.h"
 #include "gl2ps-renderer.h"
 #include "graphics.h"
diff --git a/src/gl-render.h b/src/gl-render.h
--- a/src/gl-render.h
+++ b/src/gl-render.h
@@ -19,21 +19,18 @@ along with Octave; see the file COPYING.
 <http://www.gnu.org/licenses/>.
 
 */
 
 #if !defined (gl_render_h)
 #define gl_render_h 1
 
 #ifdef HAVE_WINDOWS_H
+#define WIN32_LEAN_AND_MEAN
 #include <windows.h>
-#ifdef max
-# undef min
-# undef max
-#endif
 #endif
 
 #ifdef HAVE_GL_GL_H
 #include <GL/gl.h>
 #elif defined HAVE_OPENGL_GL_H || defined HAVE_FRAMEWORK_OPENGL
 #include <OpenGL/gl.h>
 #endif
 
